python/datasources/cawp.py

```diff
@@ -0,0 +1,316 @@
+STATE_COL_CAWP_TOTALS = "State"
+STATE_COL_LINE = "state"
+POSITION_COL = "position"
+
+# PROPUB COLUMNS
+IN_OFFICE_COL = "in_office"
+
+# CAWP CONSTS
+NATIONAL = "national"
+STATE = "state"
+
+
+CAWP_RACE_GROUPS_TO_STANDARD = {
+    'Asian American/Pacific Islander': Race.ASIAN_PAC.value,
+    'Latina': Race.HISP.value,
+    'Middle Eastern/North African': Race.MENA.value,
+    # MULTI = "Multiracial Alone" + women w multiple specific races
+    'Multiracial Alone': Race.MULTI_OR_OTHER_STANDARD.value,
+    'Native American/Alaska Native/Native Hawaiian': Race.AIANNH.value,
+    'Black': Race.BLACK.value,
+    'White': Race.WHITE.value,
+    'Unavailable': Race.UNKNOWN.value,
+    'All': Race.ALL.value
+}
+
+CAWP_DATA_TYPES = {
+    STATE: ["Territorial/D.C. Senator", "Territorial/D.C. Representative", "State Representative", "State Senator"],
+    NATIONAL: ["U.S. Representative", "U.S. Senator", "U.S. Delegate"]
+}
+
+
+def get_pct(numerator: int, denominator: int):
+    """ Prevents division by zero errors from using util pct function
+    numerator: top number as int
+    denominator: bottom number as int
+
+    returns: float representing the % value of that rounded fraction
+
+    Note: Division by zero just returns 0.0 instead of error
+    """
+    if denominator == 0:
+        return 0.0
+    return percent_avoid_rounding_to_zero(numerator, denominator)
+
+
+def get_standard_code_from_cawp_phrase(cawp_place_phrase: str):
+    """ Accepts a CAWP place phrase found in the LINE ITEM table
+    `{STATE_COL_LINE NAME} - {CODE}` with the standard 2 letter code
+     """
+
+    # swap out non-standard 2 letter codes
+    cawp_place_phrase = {"American Samoa - AM":
+                         "American Samoa - AS",
+                         "Northern Mariana Islands - MI":
+                         "Northern Mariana Islands - MP"}.get(
+                             cawp_place_phrase, cawp_place_phrase)
+
+    # only keep 2 letter code portion
+    place_terms_list = cawp_place_phrase.split(" - ")
+    place_code = place_terms_list[1]
+
+    return place_code
+
+
+def remove_markup(datum: str):
+    """Returns the string with any asterisks and/r italics markup removed """
+    datum = str(datum)
+    return datum.replace("<i>", "").replace("</i>", "").replace("*", "")
+
+
+def count_matching_rows(df, place_code: str, gov_level: str, race_to_match: str):
+    """ Accepts a dataframe, a level of government, a place code,
+    and a race name (CAWP terminology) string to match within the
+     race_ethnicity column. It then counts the number of
+    rows where those conditions are all met  """
+
+    df = df[(df[POSITION_COL].isin(CAWP_DATA_TYPES[gov_level]))]
+
+    # to get national values, don't restrict by state
+    if place_code != constants.US_ABBR:
+        df = df[(
+            df[POSTAL_COL] == place_code)]
+
+    # to get ALL women, don't restrict by race
+    if race_to_match == std_col.ALL_VALUE:
+        race_to_match = ""
```

Comment:
hmm ya it's kinda weird. I adjusted it; i think it's better? It's strange because I don't want to restrict the DF if it's "multi" because we need to include "other" and multiple listed races with MULTI. Let me know if this change makes more sense

Reply:
Yeah,  I think your new solution is good